pallets/data-preservers/src/lib.rs

```diff
@@ -62,36 +61,243 @@ use {
+    }
+
+    impl<T: Config> ParaIdsFilter<T> {
+        #[allow(clippy::len_without_is_empty)]
+        pub fn len(&self) -> usize {
+            match self {
+                Self::AnyParaId => 0,
+                Self::Whitelist(list) | Self::Blacklist(list) => list.len(),
+            }
+        }
+
+        pub fn can_assign(&self, para_id: &ParaId) -> bool {
+            match self {
+                ParaIdsFilter::AnyParaId => true,
+                ParaIdsFilter::Whitelist(list) => list.contains(para_id),
+                ParaIdsFilter::Blacklist(list) => !list.contains(para_id),
+            }
+        }
+    }
+
+    #[derive(
+        RuntimeDebug, PartialEq, Eq, Encode, Decode, Clone, TypeInfo, Serialize, Deserialize,
+    )]
+    pub enum ProfileMode {
+        Bootnode,
+        Rpc { supports_ethereum_rpcs: bool },
+    }
+
+    /// Profile with additional data:
+    /// - the account id which created (and manage) the profile
+    /// - the amount deposited to register the profile
+    #[derive(
+        RuntimeDebugNoBound,
+        PartialEqNoBound,
+        EqNoBound,
+        Encode,
+        Decode,
+        CloneNoBound,
+        TypeInfo,
+        Serialize,
+        Deserialize,
+    )]
+    #[scale_info(skip_type_params(T))]
+    pub struct RegisteredProfile<T: Config> {
+        pub account: T::AccountId,
+        pub deposit: BalanceOf<T>,
+        pub profile: Profile<T>,
+        /// There can be at most 1 assignment per profile.
+        pub assignment: Option<(ParaId, AssignmentWitnessOf<T>)>,
+    }
+
+    /// Computes the deposit cost of a profile.
+    pub trait ProfileDeposit<Profile, Balance> {
+        fn profile_deposit(profile: &Profile) -> Result<Balance, DispatchErrorWithPostInfo>;
+    }
+
+    /// Implementation of `ProfileDeposit` based on the size of the SCALE-encoding.
+    pub struct BytesProfileDeposit<BaseCost, ByteCost>(PhantomData<(BaseCost, ByteCost)>);
+
+    impl<Profile, Balance, BaseCost, ByteCost> ProfileDeposit<Profile, Balance>
+        for BytesProfileDeposit<BaseCost, ByteCost>
+    where
+        BaseCost: Get<Balance>,
+        ByteCost: Get<Balance>,
+        Profile: Encode,
+        Balance: TryFrom<usize> + CheckedAdd + CheckedMul,
+    {
+        fn profile_deposit(profile: &Profile) -> Result<Balance, DispatchErrorWithPostInfo> {
+            let base = BaseCost::get();
+            let byte = ByteCost::get();
+            let size: Balance = profile
+                .encoded_size()
+                .try_into()
+                .map_err(|_| ArithmeticError::Overflow)?;
+
+            let deposit = byte
+                .checked_mul(&size)
+                .ok_or(ArithmeticError::Overflow)?
+                .checked_add(&base)
+                .ok_or(ArithmeticError::Overflow)?;
+
+            Ok(deposit)
+        }
+    }
+
+    /// Allows to process various kinds of payment options for assignments.
+    pub trait AssignmentPayment<AccountId> {
```

Comment:
Can we move it to a assignment.rs file within this crate? so that the pallet becomes a bit cleaner